### ISSUE TITLE
extract a single snapshot from tar.gz archive

### ISSUE THREAD

**Issue by kntasks** [bit-team/backintime#1208]
I'm using tar.gz to store the backintime repository in a single file then copying that tar.gz file to an external drive for safe keeping. 

Need to restore just one snapshot, so, I copy the tar.gz file from external hard drive to the local disk on laptop then attempt to just extract that particular snapshot using this command:

`tar -xzvpf backintime-void-hp-gaming-211208.tar.gz backintime/k-void/root/1/20211208-100001-237`

However, after seeing a few files extract successfully, the terminal output shows the following:

```
tar: backintime/k-void/root/1/20211208-100001-237/backup/var/lib/flatpak/runtime/org.kde.Platform.Locale/x86_64/5.14/b58c93affee647ea545bdfdaabcb74a7639e4aa0692831ad902e960e5708c736/files/ca/lib/ca_ES@euro/LC_NUMERIC: Cannot hard link to ‘backintime/k-void/root/1/20211207-225756-339/backup/var/lib/flatpak/runtime/org.kde.Platform.Locale/x86_64/5.14/b58c93affee647ea545bdfdaabcb74a7639e4aa0692831ad902e960e5708c736/files/it/lib/it_IT@euro/LC_NUMERIC’: No such file or directory
backintime/k-void/root/1/20211208-100001-237/backup/var/lib/flatpak/runtime/org.kde.Platform.Locale/x86_64/5.14/b58c93affee647ea545bdfdaabcb74a7639e4aa0692831ad902e960e5708c736/files/ca/lib/ca_ES@euro/LC_TIME
tar: backintime/k-void/root/1/20211208-100001-237/backup/var/lib/flatpak/runtime/org.kde.Platform.Locale/x86_64/5.14/b58c93affee647ea545bdfdaabcb74a7639e4aa0692831ad902e960e5708c736/files/ca/lib/ca_ES@euro/LC_TIME: Cannot hard link to ‘backintime/k-void/root/1/20211207-225756-339/backup/var/lib/flatpak/runtime/org.kde.Platform.Locale/x86_64/5.14/b58c93affee647ea545bdfdaabcb74a7639e4aa0692831ad902e960e5708c736/files/ca/lib/ca_ES@euro/LC_TIME’: No such file or directory
backintime/k-void/root/1/20211208-100001-237/backup/var/lib/flatpak/runtime/org.kde.Platform.Locale/x86_64/5.14/b58c93affee647ea545bdfdaabcb74a7639e4aa0692831ad902e960e5708c736/files/ca/lib/ca_ES@euro/LC_ADDRESS
```

There are several of these errors and inevitably the process finishes with:

`tar: Exiting with failure status due to previous errors`

Is there any way to store the backintime repository as an archive (tar, rar, zip etc) then extract a single snapshot?

I only ask because extracting the whole tar archive (around 20GB in my case) takes over 1 day to do. The only time I've been able to extract successfully is, as I said when the complete tar archive is extracted or when the oldest snapshot is extracted. Anything else fails with errors such as the above.

**Comment by gsker** [comment]
What version of tar? And what is the command you use to create the tar file?
I've not done this but I would expect if I did I would need to use this option:
* tar --version
     `tar (GNU tar) 1.34`
* --hard-dereference
              Follow hard links; archive and dump the files they refer

**Comment by kntasks** [comment]
> What version of tar? And what is the command you use to create the tar file? I've not done this but I would expect if I did I would need to use this option:
> 
> * tar --version
>   `tar (GNU tar) 1.34`
> * --hard-dereference
>   Follow hard links; archive and dump the files they refer

Thanks for your suggestion. 

My tar (GNU tar) is 1.34. 

backintime repository (directory) is in directory /media/data/backup

So I change directory to it then ran: 

1) `tar -czvpf nameoftarfile.tar.gz backintime  `

this produced a tar.gz file of about 5GB. I think there was only about 2 snapshots.

Then I tried extracting the tar.gz file and it took forever. Assuming the long extraction time was due to the compression used, I tried this:

2) `tar -cpvf nameoftarfile.tar backintime `

That one had several snapshots (because after backintime installation I had configured scheduled backups of 1 every 2 hours). There were about 10 snapshots. Extracting took over 1 day.

So, next, I tried extracting the oldest snapshot within the tar archive. That completed without errors and didn't take very long. 

Next, I attempted to extract a snapshot done a few days after the oldest snapshot. That extract failed with errors such as what I mentioned in the original post.

I've now tried your suggestion, so modified my initial tar command to this:

3)` tar --hard-dereference -czvpf /media/data1/archive/backintime-211231-2152.tar.gz backintime`

Ran this command at about 4:45 pm today and by 7 pm, it was still running. Saw the .tar.gz grow to over 54GB so did a ctrl-C to stop, as I was probably going to run out of disk space before it finished. 

I have 18 snapshots in the backintime directory. Doing `ncdu` in for one of the snapshot directories I can see it takes up about 9GB. Just cannot be handling a tar.gz file well over 50GB to copy to/from external hard drive every week. So it doesn't seem I can use the `--hard-dereference `option to backup the whole backintime directory to external hard drive. However, I think as a workaround (or next best option), I could probably use (1) to backup whole backintime directory to external hard drive, then use (3) to just backup the last snapshot. That way, if laptop does suffer a hard drive failure or similar, I can replace hard drive then extract last snapshot and go from there. If I need a particular snapshot, then I guess, I would need to leave laptop running overnight and extract the whole backintime repository (directory) stored in nameoftarfile.tar.gz. i.e. the tar.gz archive made using(1).

If you or anyone else have other ideas and suggestions please revert

**Comment by gsker** [comment]
Fascinating.  I dug into this a bunch. There's no out-of-the-box way to do it that I found. I tried tricks with tar incremental backups and --newer-mtime and tried `duplicity` and then `rdup`.  I did not think it would be so crazy hard. 

I wonder if it would work to have a second backup in backintime that's just for the external drive.  
The compression might be a big deal for you though.

**Comment by kntasks** [comment]
> Fascinating. I dug into this a bunch. There's no out-of-the-box way to do it that I found. I tried tricks with tar incremental backups and --newer-mtime and tried `duplicity` and then `rdup`. I did not think it would be so crazy hard.
> 
> I wonder if it would work to have a second backup in backintime that's just for the external drive. The compression might be a big deal for you though.

I changed tack and am now using rsync to copy the whole backintime repository (directory) to the external hard drive, then for the restore, mounting that external drive and going in to 'settings' within backintime and then doing 'restore config' and navigating to the specific snapshot held on the external drive. I tested it on a VirtualBox machine and it worked perfectly. 

To rsync copy the backintime repository (about 11GB - not quite sure why I seem to have a .tar archive from before that is 19.7GB. I have couple of tar.gz archives which are just under 5 GB) using this command: 

`sudo rsync -avhH --progress /media/data/backup/backintime /media/veracrypt1/archive`

it took 26 minutes.

Then to restore using backintime (GUI version) from external drive it took even less. 

To be fair, the tar.gz method may still be okay if I was copying the tar.gz file to a laptop hard drive (ideally a SSD drive) then doing the extracting. If started in the morning I reckon it could complete by night (rather than leaving overnight). I think it was taking extremely long in my case because I was doing the extracting from local laptop hard drive (spindle type - as opposed to SSD) to virtual hard drive (VirtualBox virtual disk was also on same drive). The virtualbox machine I was running this on had 3 of the 8 processors of my laptop and 4GB of the 13GB memory.

I tested the extraction on laptop hard drive (spindle type - as opposed to SSD) from one partition to another (rather than doing it on a virtual machine). Started it at 2:30pm, by 1:00am it was still extracting but it was nearly done. Perhaps another 2 hours or so to complete. I terminated it, and tried the rsync method the next day.

On hindsight I should have said on the OP that if the compressed form was not possible, rsync copy was okay too. Thanks for the prompt replies and the great attempt.
